byterun/fail.c

```diff
@@ -84,22 +84,46 @@ CAMLexport void caml_raise_with_string(value tag, char const *msg)
 /* PR#5115: Failure and Invalid_argument can be triggered by
    input_value while reading the initial value of [caml_global_data]. */
 
-CAMLexport void caml_failwith (char const *msg)
+static inline value caml_get_failwith_tag (char const *msg)
 {
   if (caml_global_data == 0) {
     fprintf(stderr, "Fatal error: exception Failure(\"%s\")\n", msg);
     exit(2);
   }
-  caml_raise_with_string(Field(caml_global_data, FAILURE_EXN), msg);
+  return Field(caml_global_data, FAILURE_EXN);
 }
 
-CAMLexport void caml_invalid_argument (char const *msg)
+CAMLexport void caml_failwith (char const *msg)
+{
+  caml_raise_with_string(caml_get_failwith_tag(msg), msg);
+}
+
+CAMLexport void caml_failwith_value (value msg)
+{
+  CAMLparam1(msg);
+  caml_raise_with_arg(caml_get_failwith_tag(String_val(msg)), msg);
```

Comment:
I recommend removing `CAMLparam1` (and finding something else to say "no return").  It gives a false sense of security.  At the moment the presence of this declaration implies that it's guarding against the risk of a GC, which would presumably happen inside `caml_get_failwith_tag`, if it were changed in the future.  However if you expect a side effect from that function, then the code is implicitly relying on the order of evaluation of the arguments of `caml_raise_with_arg`, which I think is undefined behaviour.  (If the second argument was evaluated first and that function did perform a GC then it's wrong.)